### PR TITLE
Zlevel Overhaul Phase 1

### PR DESCRIPTION
**Pull request by NanakoAC**
Introduces a new zlevel system based on level datums (one per zlevel)
Also introduces the concept of a scene, which is a group of connected levels. One scene per ship, planet, asteroid, etc

Fixes and tweaks a whole ton of things to use this new system, many more changes will be needed in future though

At present, expeditions and overmap travel may not work correctly, they'll be fixed in a future phase

### CODE REVIEW

File: code/modules/multiz/level.dm [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+		return pick(SSmapping.empty_levels)
+
+	return null
+
+/proc/create_empty_level()
+	world.maxz++ //This adds a new level onto the end of the list
+	SSmapping.map_index++ //Increment the index
+	var/datum/level/newlevel = new()
+	newlevel.z = world.maxz
+	world << "Creating new level [newlevel.z]"
```
Comment (Wesreidau) [inline code/modules/multiz/level.dm]: Comment?

File: code/modules/overmap/spacetravel.dm [hunk 2]
```diff
@@ -50,6 +51,12 @@ proc/get_deepspace(x,y)
 /mob/lost_in_space()
 	return isnull(client)
 
+
+//Oddly, this proc is only for drifting mobs. shuttles don't use it
+//TODO Here:
```
Comment (Wesreidau) [inline code/modules/overmap/spacetravel.dm:56]: TODO Line

File: code/modules/overmap/spacetravel.dm [hunk 3]
```diff
@@ -67,6 +74,8 @@ proc/overmap_spacetravel(var/turf/space/T, var/atom/movable/A)
 	var/ny = 1
 	var/nz = 1
 
+	//Figure out what edge of the world we're off
+		//Todo: Replace this shit
```
Comment (Wesreidau) [inline code/modules/overmap/spacetravel.dm]: Have you replaced this shit? Also, recommending we use professional verbage.

File: code/modules/overmap/spacetravel.dm [hunk 4]
```diff
@@ -85,14 +94,24 @@ proc/overmap_spacetravel(var/turf/space/T, var/atom/movable/A)
 
 	testing("[A] spacemoving from [M] ([M.x], [M.y]).")
 
+
+	//This level is where the overmap itself exists. the thing which is visible in helm consoles
 	var/turf/map = locate(M.x,M.y,GLOB.using_map.overmap_z)
 	var/obj/effect/overmap/TM
 	for(var/obj/effect/overmap/O in map)
+		//This is finding all the possible visitable locations within the specified map tile
+		//TODO: Remove this stupid 50% probabiltiy
```
Comment (Wesreidau) [inline code/modules/overmap/spacetravel.dm]: Is this complete? Also verbage.

File: code/modules/overmap/spacetravel.dm [hunk 5]
```diff
@@ -85,14 +94,24 @@ proc/overmap_spacetravel(var/turf/space/T, var/atom/movable/A)
...
 		if(O != M && O.in_space && prob(50))
 			TM = O
 			break
+
 	if(!TM)
+		//This function gets an empty level.
+			//Todo: Make the returned level be removed from the empty levels list and be named appropriately with overmap grid goords
```
Comment (Wesreidau) [inline code/modules/overmap/spacetravel.dm:110]: Complete?

File: code/controllers/subsystems/mapping.dm [hunk 6]
```diff
@@ -68,3 +89,22 @@ SUBSYSTEM_DEF(mapping)
...
+//We do this by looping through numbers until we fail to retrieve a tile
+/hook/roundstart/proc/check_zlevel_datums()
+	var/done = FALSE
+	var/index = 1
+	while(!done)
+		var/turf/T = locate(1,1,index)
+		if (istype(T))//If we found T then this zlevel exists
+			var/datum/level/L = SSmapping.zlevels["[T.z]"]
+			if (!istype(L))
+				log_world("Error: Z level [T.z] has no associated level datum.")
```
Comment (Wesreidau) [inline code/controllers/subsystems/mapping.dm:104]: Debug message.

Comment (NanakoAC) [reply code/controllers/subsystems/mapping.dm:104]: this one is intentional, it should be left in permanantly.
It warns map makers that they've screwed up by not placing a level landmark

File: code/modules/overmap/spacetravel.dm (hunk 2 above)
Comment (NanakoAC) [reply code/modules/overmap/spacetravel.dm:56]: later, once we work on expeditions

File: code/modules/overmap/spacetravel.dm (hunk 3 above)
Comment (NanakoAC) [reply code/modules/overmap/spacetravel.dm]: all todos in this file are for phase 3

**Pull request merged**